go/mysql/binlog_event_compression.go

```diff
@@ -193,77 +208,149 @@ func (tp *TransactionPayload) read(data []byte) error {
 }
 
 // decode decompresses the payload and extracts the internal binlog
-// events.
+// events, assigning the iterator to a function that can then be used
+// to retrieve the events from the decompressed transaction one by one.
 func (tp *TransactionPayload) decode() error {
-	if tp.CompressionType != TransactionPayloadCompressionZstd {
+	if tp.compressionType != TransactionPayloadCompressionZstd {
 		return vterrors.New(vtrpcpb.Code_INTERNAL,
-			fmt.Sprintf("TransactionPayload has unsupported compression type of %d", tp.CompressionType))
+			fmt.Sprintf("TransactionPayload has unsupported compression type of %d", tp.compressionType))
+	}
+
+	decompressedPayload, decompressedFile, err := tp.decompress()
+	if decompressedFile != nil && len(decompressedPayload) > 0 { // Should never happen
+		return vterrors.Errorf(vtrpcpb.Code_INTERNAL,
+			"[BUG] error uncompressing transaction payload: illegal mix of in-memory and file-based decompression")
+	}
+
+	// Processing large payloads via a temporary file.
```

Comment:
We coud also use memory-mapped files so that we don't have two different code paths for decoding the events: `https://pkg.go.dev/golang.org/x/exp/mmap` and minimise the IO related operations. It should also give us better performance.  

We would need a panic handler in the iterator, since IO errors during paging would result in accessing invalid memory regions.

Reply:
That's a good idea. Definitely worth exploring and I'll do that. Thanks!

Reply:
I moved to using mmap here: https://github.com/vitessio/vitess/pull/16328/commits/153463fb154cb00c5245ed0d6ac209ab0a5b29f6

I'm not sure that it's any faster or more efficient today (was not in my testing), but:
1. It's a simpler and cleaner internal implementation across the two methods (in-memory and temp file)
2. It should make it easier for us to add optimizations in the future such as multiple readers to share the decompressed payload temp file across streams/vstreamers on a given vttablet.

Thanks for the great idea/suggestion!